### PR TITLE
feat(parquet): prune row groups before reading.

### PR DESCRIPTION
**Pull request by RinChanNOWWW**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Prune row groups before reading.

If there are column statistics in the parquet file, we can do row group pruning by these statistics.
Otherwise, we read the whole row groups.

### CODE REVIEW

File: src/query/storages/parquet/src/parquet_reader/meta.rs [hunk 1]
```diff
@@ -43,4 +52,80 @@ impl ParquetReader {
...
+                ));
+            }
+
+            let field = &schema.fields[*index];
+            let column_stats = pread::statistics::deserialize(field, rgs)?;
+            stats_of_row_groups.insert(*index, BatchStatistics::from(column_stats));
+        }
+
+        for (rg_idx, _) in rgs.iter().enumerate() {
+            let mut cols_stats = HashMap::new();
```
Comment (Xuanwo) [inline src/query/storages/parquet/src/parquet_reader/meta.rs:83]: How about `HashMap::with_capacity(stats.capacity())`

File: src/query/storages/pruner/src/range_pruner.rs [hunk 2]
```diff
@@ -55,6 +55,9 @@ impl RangePruner for RangeFilter {
     }
 }
 
+/// Create a new [`RangePruner`] from expression and schema.
```
Comment (Xuanwo) [inline src/query/storages/pruner/src/range_pruner.rs:58]: I prefer to have a `RangePruner::new()`

File: src/query/storages/fuse/fuse/src/pruning/pruning_executor.rs [hunk 3]
```diff
@@ -79,7 +79,7 @@ impl BlockPruner {
             .and_then(|p| p.limit);
 
         // prepare the limiter. in case that limit is none, an unlimited limiter will be returned
-        let limiter = limiter::new_limiter(limit);
+        let limiter = limiter_pruner::new_limiter(limit);
```
Comment (Xuanwo) [inline src/query/storages/fuse/fuse/src/pruning/pruning_executor.rs:82]: I prefer to have a `LimiterPruner::new()` API which carries more context about this struct. `new_limiter` could be anything if imported directly.

File: src/query/storages/pruner/src/range_pruner.rs (hunk 2 above)
Comment (RinChanNOWWW) [reply src/query/storages/pruner/src/range_pruner.rs:58]: But `RangePruner` is a trait.

File: src/query/storages/fuse/fuse/src/pruning/pruning_executor.rs (hunk 3 above)
Comment (RinChanNOWWW) [reply src/query/storages/fuse/fuse/src/pruning/pruning_executor.rs:82]: `LimiterPruner` is also a trait.

**Review by Xuanwo — APPROVED** [review]

**Comment by dantengsky** [comment]
looks good to me. It will be better if there were unit or stateless tests

**Pull request merged**